src/Processors/QueryPlan/Optimizations/convertJoinToIn.cpp

```diff
@@ -0,0 +1,169 @@
+    extern const SettingsUInt64 use_index_for_in_with_subqueries_max_values;
+}
+
+namespace DB::QueryPlanOptimizations
+{
+
+bool findReadingStep(QueryPlan::Node & node)
+{
+    IQueryPlanStep * step = node.step.get();
+    if (dynamic_cast<ISourceStep *>(step))
+    {
+        /// might be more cases here
+        if (typeid_cast<ReadFromSystemColumns *>(step) || typeid_cast<ReadFromSystemReplicas *>(step))
+            return false;
+        return true;
+    }
+
+    if (node.children.size() != 1)
+        return false;
+
+    if (typeid_cast<ExpressionStep *>(step) || typeid_cast<FilterStep *>(step) || typeid_cast<ArrayJoinStep *>(step))
+        return findReadingStep(*node.children.front());
+
+    if (auto * distinct = typeid_cast<DistinctStep *>(step); distinct && distinct->isPreliminary())
+        return findReadingStep(*node.children.front());
+
+    return false;
+}
+
+size_t tryConvertJoinToIn(QueryPlan::Node * parent_node, QueryPlan::Nodes & nodes, const Optimization::ExtraSettings & /*settings*/)
+{
+    auto & parent = parent_node->step;
+    auto * join = typeid_cast<JoinStepLogical *>(parent.get());
+    if (!join)
+        return 0;
+    auto & join_info = join->getJoinInfo();
+    if (join_info.strictness != JoinStrictness::All)
+        return 0;
+    if (!isInner(join_info.kind))
+        return 0;
+    if (join->getJoinSettings().join_use_nulls)
+        return 0;
+
+    const auto & left_input_header = join->getInputHeaders().front();
+    const auto & right_input_header = join->getInputHeaders().back();
+    const auto & output_header = join->getOutputHeader();
+
+    bool left = false;
+    bool right = false;
+    for (const auto & column_with_type_and_name : output_header)
+    {
+        left |= left_input_header.has(column_with_type_and_name.name);
+        right |= right_input_header.has(column_with_type_and_name.name);
+    }
+    /// All come from left side?
+    if (!(left && !right))
+        return 0;
+
+    if (join_info.expression.condition.predicates.empty())
+        return 0;
+
+    Header left_predicate_header;
+    Header right_predicate_header;
+    /// column in predicate is null, so get column here
+    for (const auto & predicate : join_info.expression.condition.predicates)
+    {
+        const Header & left_header = parent_node->children.front()->step->getOutputHeader();
+        const auto * left_column = left_header.findByName(predicate.left_node.getColumn().name);
+        if (!left_column)
+            return 0;
+        left_predicate_header.insert(*left_column);
+
+        const Header & right_header = parent_node->children.back()->step->getOutputHeader();
+        const auto * right_column = right_header.findByName(predicate.right_node.getColumn().name);
+        if (!right_column)
+            return 0;
+        right_predicate_header.insert(*right_column);
+    }
+
+    /// Steps like ReadFromSystemColumns would build set in place, different from here, so return.
```

Comment:
ReadFromMergeTree also could build the set inplace.
That's fine so far. We can set better limits in the future.